### PR TITLE
as nvim-completion-manager source.

### PR DESCRIPTION
**Pull request by roxma**
Add [nvim-completion-manager](https://github.com/roxma/nvim-completion-manager) source, async support, parallel execution that will never block the ui.

### CODE REVIEW

**Comment by roxma** [comment]
```vim
let g:clang_make_default_keymappings = 0
```

in order to test it with NCM.

By the way, I think `g:clang_make_default_keymappings` should be 0 by defualt to better work with other framework.

**Comment by roxma** [comment]
Benefits gained from this PR:

- real async
- no need to map `.`, `>` key from clang_plugin
- won't miss the popup menu if you type more characters before the completion calculation is finished.

Origional behavior:

clang


With NCM:

ncm_clang

**Comment by roxma** [comment]
Extra benifit, just for fun, support markdown:

nvim_clang_markdown

**Comment by roxma** [comment]
Anyone helps reviewing the code?  This PR does not break the already working default behavior.

@xaizek

**Comment by xaizek** [comment]
I'm not sure why you would want to include these changes in clang_complete. It looks like they use libclang bindings and some of plugin configuration, but not core part of the plugin.

Also I'm not using it for a long time now, not using nvim and thus reluctant to add changes that I can't even test properly. Pretty much the only reason why I pulled python3 support recently is because people were waiting for it for years. By the way, you limit it to python2 while both versions should work now.

**Comment by roxma** [comment]
@xaizek 

> I'm not sure why you would want to include these changes in clang_complete.  It looks like they use libclang bindings and some of plugin configuration, but not core part of the plugin.

the `.clang_complete` style configuration looks nice, so I would like it to be integrated in the same plugin, instead of spliting a fork.

> . Pretty much the only reason why I pulled python3 support recently is because people were waiting for it for years. By the way, you limit it to python2 while both versions should work now.

The official clang python binding only supports python2, see this: https://github.com/llvm-mirror/clang/commit/abdad67b94ad4dad2d655d48ff5f81d6ccf3852e

And don't worry about the python file added by this PR. I would be executed as a standalone process, if [nvim-completion-manager](https://github.com/roxma/nvim-completion-manager) is installed , so It will not affect clang_complete's default behavior, even if I limit it to use python2.

**Comment by xaizek** [comment]
> The official clang python binding only supports python2

Version here is different from the official one and works with both python2 and python3 in my limited tests and for other people. Some string encoding was added and even though it's not universal, it should work unless you use non-ascii characters to name things.

File: plugin/libclang.py [hunk 1]
```diff
@@ -64,10 +65,31 @@ def getBuiltinHeaderPath(library_path):
...
-def initClangComplete(clang_complete_flags, clang_compilation_database, \
-                      library_path):
+def initClangComplete():
+
+  # import snippets
+  try:
+    # m = importlib.import_module(modulename)
+    # for v in m:
+    #   # globals[
+    exec('import snippets.' + vim.eval('g:clang_snippets_engine') + ' as current_snippets')
```
Comment (xaizek) [inline plugin/libclang.py]: I guess this line doesn't bring `current_snippets` into function scope in python3. Python complains about unknown `current_snippets` variable. Also there is commented out code above.

**Comment by roxma** [comment]
Hi, I decided to maintain a fork focusing on integration with nvim-completion-manager, some unnecessary features will be removed for simplicity, such as builtin snippet engine.

Thanks for the great work. Please don't feel offended.

**Pull request closed without merging**